spec/index.adoc

```diff
@@ -3996,9 +4021,93 @@ S with
+
+We define an auxiliary method that handles calls from composite query methods by performing a call graph traversal. It can also be (trivially) invoked for query methods that do not make further calls.
+....
+composite_query_helper(S, Cycles, Depth, Caller, Canister_id, Method_name, Arg) =
+  let Mod = S.canisters[Canister_id].module
+  let F = Mod.composite_query_methods[Method_name]
+  let Cert <- { Cert | verify_cert(Cert) and
+                       lookup(["canister", Canister_id, "certified_data"], Cert) = Found S.certified_data[Canister_id] and
+                       lookup(["time"], Cert) = Found S.system_time // or “recent enough”
+              }
+  let Env = { time = S.time[Canister_id];
+              global_timer = S.global_timer[Canister_id];
+              balance = S.balances[Canister_id];
+              freezing_limit = freezing_limit(S, Canister_id);
+              certificate = Cert;
+              status = simple_status(S.canister_status[Canister_id]);
+              canister_version = S.canister_version[Canister_id];
+            }
+  if S.canisters[Canister_id] ≠ EmptyCanister and
+     S.canister_status[Canister_id] = Running and
+     S.balances[Canister_id] >= freezing_limit(S, Canister_id) and
+     Method_name ∈ dom(Mod.composite_query_methods) and
+     Cycles >= MAX_CYCLES_PER_MESSAGE
+  then
+     let W = S.canisters[Canister_id].wasm_state
+     let R = F(Arg, Caller, Env)(W)
+     if R = Trap trap
+     then Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles - trap.cycles_used)
+     else if R = Return {new_state = W'; new_calls = Calls; response = Response; cycles_used = Cycles_used}
+     then
+        W := W'
+        if Cycles_used > MAX_CYCLES_PER_MESSAGE
+        then
+           Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles - MAX_CYCLES_PER_MESSAGE) // single message execution out of cycles
+        Cycles := Cycles - Cycles_used
+        if Response = NoResponse
+        then
+           while Calls ≠ []
+           do
+              if Depth = MAX_CALL_DEPTH_COMPOSITE_QUERY
+              then
+                 Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // max call graph depth exceeded
+              let Calls' · Call · Calls''  = Calls
+              Calls := Calls' · Calls''
+              if S.canister_subnet[Canister_id] ≠ S.canister_subnet[Call.callee]
+              then
+                 Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // calling to another subnet
+              let (Response', Cycles_used') = composite_query_helper(S, Cycles, Depth + 1, Canister_id, Call.callee, Call.method_name, Call.arg)
+              Cycles := Cycles - Cycles_used'
+              if Cycles < MAX_CYCLES_PER_RESPONSE
+              then
+                 Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // composite query out of cycles
+              let F' = Mod.callbacks(Call.callback, Response', 0, Env, 0)
```

Comment:
Given that
```
ic0.call_cycles_add : (amount : i64) -> ();                                 // U Ry Rt T
ic0.call_cycles_add128 : (amount_high : i64, amount_low: i64) -> ();        // U Ry Rt T
```
which also matches the implementation I thought that no cycles can ever be attached in a composite query execution.

But they could be actually attached to calls made from callbacks (Ry, Rt). We can also use the following API in callbacks:
```
ic0.certified_data_set
ic0.global_timer_set
ic0.msg_cycles_accept
ic0.msg_cycles_accept128
```

I believe that we should introduce a pair of new modes, e.g., CRy and CRt, to denote callbacks in composite query contexts and make the above API unavailable in CRy and CRt. What do you think?

The alternative (your second option) could be confusing as we would allow sending cycles but we would never actually commit cycles sent in such a way. Moreover, we would then also need to allow setting certified variables and global timer although there would be no practical effect of that (except for retrieving the new certified variables in the same query - without signatures - and retrieving the previous value of global timer by setting a new value that is never gonna be used).

@ulan

Reply:
This suggests that refunded cycles are set to zero for callbacks in composite queries: https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/execution_environment/src/query_handler/query_context.rs?L411

Reply:
> I believe that we should introduce a pair of new modes, e.g., CRy and CRt, to denote callbacks in composite query contexts and make the above API unavailable in CRy and CRt. What do you think?

I agree. This seems to be the cleanest solution.

Reply:
I've added the new modes (and also one for callback cleanup in composite queries).